### DISCUSSION TITLE
Compute correlation every second in a FileDevice

### DISCUSSION THREAD

**Discussion by MikelElorza** [PicoQuant/snAPI#38] [Q&A] [answered]
I would like to compute the G2 curve of a file every second (G2 curve of the photons arriving on the first second, G2 curve for the photons arriving in the second second and so on...) instead of computing the G2 using all the data of the file. Is there any way to select the photons you want to analyse and compute their correlations (using the correlation class)? 
I was trying to measure just one second and get the correlation iteratively:
-sn.correlation.measure(acqTime=1000)
-sn.getG2Data()
but acqTime is ignored in the FileDevices as far as I know.

Thank you in advance,
Mikel

**Comment by snizzleorg** [comment]
it is currently not possible to correlate only parts of a file. The use case however seems interesting and we will treat it as a feature suggestion. 
Maybe you can add a few more details about the application so that also other people can weight in. This would help us to assess demand but also scope of such a feature.

**Reply by MikelElorza** [reply]
I have some quantum dots suspended in a liquid and I measure their emission using a confocal microscope. As the quantum dots enter and leave the confocal volume I get photon bursts. I would like to analyse these bursts one by one. Currently I am unfolding the T2 file, subdividing the data and processing (getting correlations, time traces etc.)  each part independently (using python scripts written by myself). 
It would be extremely helpful for this application to be able to only read parts of a file and analyse it using snAPI. Something like this would be ideal in my case:
- sn = snAPI()
- sn.getFileDevice(filename)
- sn.correlation.measure(from  1ms to 3ms)
- g2,lags=sn.correlation.getG2Data()
- sn.unfold.measure(from  1ms to 3ms)
- times1=getTimesByChannel(1)
The idea is to have access to a certain part of the file and to be able to use snAPI classes and methods with it.

Thank you very much!

**Reply by snizzleorg** [reply]
I think for this use case also some kind of burst identification would be great, right? So basically to get all the counts of a particular burst and then correlate those counts.

**Comment by MikelElorza** [comment]
Yes, exactly! It would be great to have some sort of identification. So the idea would be to have the option to 'add' some identification to each photon in the postprocessing of the data so you can separe the data according to the identification?

**Comment by snizzleorg** [comment]
My idea would be to extend the Heralding function which allows to filter the photons.
I think it might be a good idea to do this in two steps. First step would be selecting burst start/stop times (from a time trace which would need to be calculated) then the burst start/stop times could be used as an input for the Heralding. The disadvantage would be that the same file would need to be processed twice. Once for identifying the bursts and second for the correlation.

**Reply by MikelElorza** [reply]
This idea sounds great to me!

**Comment by MikelElorza** [comment]
I have also been using markers to separate data from different spatial positions or from different experimental conditions. I have always wanted to analyse this data part by part (from marker to marker) using snAPI. I have done this reading all the data using unfold class and postprocessing it using my own scripts. I imagine that this identification might also be used to get the counts between markers.  
Thank you.

**Comment by tpoint75** [comment, marked as answer]
Hi @MikelElorza I made a new update:

just update it with
pip install --upgrade snAPI

Now we have a new function to split the calculation of the g(2) in consecutive windows:
https://picoquant.github.io/snAPI/snAPI.Main.html#snAPI.Main.Correlation.setSequenceMode

I also made a script for you that:
https://github.com/PicoQuant/snAPI/blob/main/demos/Demo_g2overTime.py
Comment line 11 and uncomment line 12 and it works with ptu-files.
It should also be fast enough, to calculate these g(2) windows in realtime. Only the pcolormesh plot can't handle this for long..

Please tell us, if this solves your problem!

Best regards,
